search-sentiments-map/routes/trends.js

```diff
@@ -1,35 +1,233 @@
+ * Example data structure for a `trendsEntry`:
+ * {timestamp: 111111111,
+    trendsByCountry: [{
+        country: US,
+        trends: [{
+          topic: Donald Trump,
+          articles: [title1, ..., title7],
+          sentimentScore: 0.2,
+          }...
+        ]}, {
+        country: UK,
+        trends: [..., ...]
+      }...
+    ],
+    globalTrends: [{
+        trendTopic: X,
+        count: 5,
+    }, ...]
+   }
+ * @param {!Array<JSON>} trendsByCountry An array where each element is a country
+ * and its trends.
+ */
+async function saveTrendsAndDeletePrevious(trendsByCountry) {
+  await deleteAncientTrend();
+
+  const trendsEntryKey = datastore.key(TRENDS_DATA_KIND);
+  const trendsEntry = {
+    key: trendsEntryKey,
+    data: {
+      timestamp: Date.now(),
+      trendsByCountry: trendsByCountry,
+      globalTrends: getGlobalTrends(trendsByCountry),
+    },
+  };
+
+  try {
+    await datastore.save(trendsEntry);
+    console.log(`TrendsEntry ${trendsEntryKey.id} created successfully.`);
+  } catch (err) {
+    console.error('ERROR:', err);
+  }
+}
+
+/** 
+ * Deletes the oldest trend record if it was saved more than 7 days ago. 
+ */
+async function deleteAncientTrend() {
+  // Query entries in ascending order of the time of creation.
+  const query = datastore.createQuery(TRENDS_DATA_KIND).order('timestamp');
+  const [trendsEntries] = await datastore.runQuery(query);
+
+  if (trendsEntries.length === 0) {
+    return;  // Nothing to delete.
+  }
+  if (Date.now() - trendsEntries[0].timestamp > 7 * 24 * 60 * 60000) {
+    const trendsEntryKey = trendsEntries[0][datastore.KEY];
+    await datastore.delete(trendsEntryKey);
+    console.log(`TrendsEntry ${trendsEntryKey.id} deleted.`);
+  }
+}
+
+/** 
+ * Deletes all trends entries in the Datastore. 
+ * Call this function when there is an update of the trends data structure. 
+ * TODO(chenyuz): Delete this function when everything is done.
+ */
+async function deleteAllTrends() {
+  const query = datastore.createQuery(TRENDS_DATA_KIND).order('timestamp');
+  const [trendsEntries] = await datastore.runQuery(query);
+
+  for (let i = 0; i < trendsEntries.length; i++) {
+    const trendKey = trendsEntries[i][datastore.KEY];
+    await datastore.delete(trendKey);
+    console.log(`Trend ${trendsEntryKey.id} deleted.`)
+  }
+}
+
+/** 
+ * Finds the trending topics that appear the most across all recorded countries 
+ * and gets the top 10 as the globally trending topics.
+ * @param {!Array<JSON>} trendsByCountry An array where each element is a country
+ * and its trends.
+ * @return {!Array<JSON>} 10 globally trending topics and the number of countries 
+ * where they are trending.
+ */
+function getGlobalTrends(trendsByCountry) {
+  // Use a map to count the number of occurences of each trend.
+  let trendCountsMap = new Map();
+  trendsByCountry.forEach(trends => {
+    trends.trends.forEach(trend => {
+      let topic = trend.topic;
```

Comment:
For the trends parameter after the first forEach, can you name it countryTrends  inseadso that the following trends.trends isn't as confusing?

Reply:
done